codegen/editable_token_from_data.php

```diff
@@ -15,419 +15,213 @@ function editable_token_from_data(
-      return new HHAST\DecimalLiteralToken($leading, $trailing, $token_text);
-    case 'octal_literal':
-      return new HHAST\OctalLiteralToken($leading, $trailing, $token_text);
-    case 'hexadecimal_literal':
-      return
-        new HHAST\HexadecimalLiteralToken($leading, $trailing, $token_text);
-    case 'binary_literal':
-      return new HHAST\BinaryLiteralToken($leading, $trailing, $token_text);
-    case 'floating_literal':
-      return new HHAST\FloatingLiteralToken($leading, $trailing, $token_text);
-    case 'execution_string':
-      return new HHAST\ExecutionStringToken($leading, $trailing, $token_text);
-    case 'single_quoted_string_literal':
-      return new HHAST\SingleQuotedStringLiteralToken(
-        $leading,
-        $trailing,
-        $token_text,
-      );
-    case 'double_quoted_string_literal':
-      return new HHAST\DoubleQuotedStringLiteralToken(
-        $leading,
-        $trailing,
-        $token_text,
-      );
-    case 'double_quoted_string_literal_head':
-      return new HHAST\DoubleQuotedStringLiteralHeadToken(
-        $leading,
-        $trailing,
-        $token_text,
-      );
-    case 'string_literal_body':
-      return new HHAST\StringLiteralBodyToken($leading, $trailing, $token_text);
-    case 'double_quoted_string_literal_tail':
-      return new HHAST\DoubleQuotedStringLiteralTailToken(
-        $leading,
-        $trailing,
-        $token_text,
-      );
-    case 'heredoc_string_literal':
-      return
-        new HHAST\HeredocStringLiteralToken($leading, $trailing, $token_text);
-    case 'heredoc_string_literal_head':
-      return new HHAST\HeredocStringLiteralHeadToken(
-        $leading,
-        $trailing,
-        $token_text,
-      );
-    case 'heredoc_string_literal_tail':
-      return new HHAST\HeredocStringLiteralTailToken(
-        $leading,
-        $trailing,
-        $token_text,
-      );
-    case 'nowdoc_string_literal':
-      return
-        new HHAST\NowdocStringLiteralToken($leading, $trailing, $token_text);
-    case 'boolean_literal':
-      return new HHAST\BooleanLiteralToken($leading, $trailing, $token_text);
-    case 'XHP_category_name':
-      return new HHAST\XHPCategoryNameToken($leading, $trailing, $token_text);
-    case 'XHP_element_name':
-      return new HHAST\XHPElementNameToken($leading, $trailing, $token_text);
-    case 'XHP_class_name':
-      return new HHAST\XHPClassNameToken($leading, $trailing, $token_text);
-    case 'XHP_string_literal':
-      return new HHAST\XHPStringLiteralToken($leading, $trailing, $token_text);
-    case 'XHP_body':
-      return new HHAST\XHPBodyToken($leading, $trailing, $token_text);
-    case 'XHP_comment':
-      return new HHAST\XHPCommentToken($leading, $trailing, $token_text);
-    case 'markup':
-      return new HHAST\MarkupToken($leading, $trailing, $token_text);
-    default:
-      throw new HHAST\UnsupportedTokenError($file, $offset, $token_kind);
+  static $class_map_with_text = dict[
```

Comment:
It should be possible to have these as class constants now

Reply:
Do you mind elaborating on what that would look like? Our in-house style still avoids classes a lot, so I'm green in that department

Reply:
```Hack
class Foo {
  const BAR = dict['a' => 'b'];
}
```

via `$codegen_class->addConst($name, $value, $nullable_comment, $renderer)`

Reply:
I originally tried doing something like this instead of using `static`, but am not sure how to type it. For the class constant I'm required to add a type, but typing as `dict<string, string>` gives `Was expecting a class, but got a string (Typing[4026])` on the line with `new $cls`. Typing as `dict<string, classname>` needs a param for the classname. Not sure what's right here?

Reply:
doing
```
class TokenClass<T super HHAST\EditableNode> {
  const dict<string, classname<T>>
    MAP = dict[
      'abstract' => HHAST\AbstractToken::class,
      'and' => HHAST\AndToken::class,
      ...
    ];
```
makes `hh_client` complain with `Was expecting a class, but got an untyped value (Typing[4026])` just like the null coalesce 🤷‍♂️